lib/eventsource.js

```diff
@@ -50,27 +51,112 @@ function EventSource (url, eventSourceInitDict) {
+    //  This will need improved so each strategy only activated after the previous is spent. One solution is to put
+    //  them all on one horrible nested block, but there should be a more readable way.
+    //  The minTimeout property of medium and long term strategies mean they will do subsequent retries correctly which
+    //  should be good enough for now
+
+    // The first retry scenario covering the first 3 seconds of retrying.
+    // We want to retry three times after the initial try to account for intermittent network issues.
+    // Exponential backoff not used here.
+    var shortTermRetryStrategy = retry.operation({
+      retries: 3,
+      factor: 0,
+      minTimeout: 1000,
+      maxTimeout: 1000,
+      randomize: false
+    })
+
+    shortTermRetryStrategy.attempt(function (currentAttempt) {
+      _emit('retrying', new Event('retrying', { shortTermRetryAttempt: currentAttempt }))
+
+      if (readyState !== EventSource.RETRYING) return
+      connect()
+      if (shortTermRetryStrategy.retry(readyState === EventSource.RETRYING)) {
+
       }
-      self.connectionInProgress = true
+    })
+
+    // Second retry scenario covering two minutes worth of retries.
+    // Wait 10 seconds before the first retry.
+    // Uses backoff and randomized timeouts.
+    var mediumTermStrategy = retry.operation({
+      retries: 15,
+      factor: 3,
+      minTimeout: 10000,
+      maxTimeout: 15000,
+      randomize: true
+    })
+
+    mediumTermStrategy.attempt(function (currentAttempt) {
+      _emit('retrying', new Event('retrying', { mediumTermRetryAttempt: currentAttempt }))
+      if (readyState !== EventSource.RETRYING) return
+      connect()
+      mediumTermStrategy.retry(readyState === EventSource.RETRYING)
+    })
+
+    // This scenario is for long term retries, which tries to connect forever, as long as the error
+    // is retryable. Uses an exponential backoff and randomizes the timeouts.
+    var longTermStrategy = retry.operation({
+      forever: true,
+      factor: 3,
+      minTimeout: 300000,
+      maxTimeout: 300000,
+      randomize: true
+    })
+
+    longTermStrategy.attempt(function (currentAttempt) {
+      _emit('retrying', new Event('retrying', { longTermRetryAttempt: currentAttempt }))
+      if (readyState !== EventSource.RETRYING) return
       connect()
-    }, self.reconnectInterval)
+      longTermStrategy.retry(readyState === EventSource.RETRYING)
+    })
+  }
+
+  function isRetryableError (error) {
+    if (error.status) {
+      var status = error.status
+      return status === 500 || status === 502 || status === 503 || status === 504
+    }
+    // Always retry on IO Errors
+    return true
+  }
+
+  function onConnectionFailed (error) {
+    if (readyState === EventSource.CLOSED) {
+      return
+    }
+
+    // If we're already in the midst of a retry then we don't want to retry again.
+    if (readyState === EventSource.RETRYING) {
```

Comment:
We shouldn't ever get here if EventSource.RETRYING == true. This is just a redundant safety check to avoid potential infinite recursion.